driver-async/src/main/com/mongodb/async/client/SingleResultCallbackSubscription.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.async.client;
+
+import com.mongodb.Block;
+import com.mongodb.MongoException;
+import com.mongodb.async.SingleResultCallback;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+class SingleResultCallbackSubscription<TResult> implements Subscription {
```

Comment:
Because the single result callback may return a list of results and we have to handle that use case.

If I split that use case out into a separate class then 95% of the code is duplicated - so it makes sense to DRY the code into a single class that coerces single non list results into a single item. 

Currently in reactive streams and rx we have two classes that do this one job with much of the code replicated.